From 3944c4dc0a4195e19f4a25cbc922eb18b3aba009 Mon Sep 17 00:00:00 2001
From: "Will.Chao" <will.chao@wnc.com.tw>
Date: Fri, 7 Jan 2022 15:27:42 +0800
Subject: [PATCH 2/2] WNC QSA63_OAM8_48 platform update

  . Modify SPI Flash partitioning and name
  . Enable NAND flash support
  . Enable USB support
  . Change PCI DMA range
---
 arch/arm64/boot/dts/marvell/ac5.dtsi        |  34 +++-
 arch/arm64/boot/dts/marvell/ac5_rd.dts      |   4 +
 arch/arm64/configs/marvell_v8_sdk_defconfig | 209 +++++++++++---------
 drivers/pci/pci.c                           |   1 +
 drivers/pci/pcie/portdrv_core.c             |   4 +-
 drivers/usb/host/ehci-orion.c               |  31 +++
 6 files changed, 178 insertions(+), 105 deletions(-)

diff --git a/arch/arm64/boot/dts/marvell/ac5.dtsi b/arch/arm64/boot/dts/marvell/ac5.dtsi
index b3eb8a94..121c2dd7 100644
--- a/arch/arm64/boot/dts/marvell/ac5.dtsi
+++ b/arch/arm64/boot/dts/marvell/ac5.dtsi
@@ -245,18 +245,23 @@
 				#size-cells = <1>;
 
 				partition@0 {
-					label = "spi_flash_part0";
-					reg = <0x0 0x800000>;
+					label = "U-Boot";
+					reg = <0x0 0x400000>;
 				};
 
 				parition@1 {
-					label = "spi_flash_part1";
-					reg = <0x800000 0x700000>;
+					label = "uboot-env";
+					reg = <0x400000 0x10000>;
+					env_size = <0x10000>;
 				};
 
 				parition@2 {
-					label = "spi_flash_part2";
-					reg = <0xF00000 0x100000>;
+					label = "onie";
+					reg = <0x410000 0x1200000>;
+				};
+				parition@3 {
+					label = "open";
+					reg = <0x1610000 0x9ffc00>;
 				};
 			};
 		};
@@ -304,8 +309,23 @@
 				label = "main-storage";
 				nand-rb = <0>;
 				nand-ecc-mode = "hw";
-				nand-ecc-strength = <12>;
+				nand-ecc-strength = <4>;
 				nand-ecc-step-size = <512>;
+				
+				partitions {
+                    compatible = "fixed-partitions";
+                    #address-cells = <1>;
+                    #size-cells = <1>;
+
+                    partition@0 {
+                        label = "boot";
+                        reg = <0 0x200000>;
+                    };
+                    partition@200000 {
+                        label = "rootfs";
+                        reg = <0x200000 0xfe00000>;
+                    };
+                };
 			};
 		};
 
diff --git a/arch/arm64/boot/dts/marvell/ac5_rd.dts b/arch/arm64/boot/dts/marvell/ac5_rd.dts
index 2892c705..8499d4cb 100644
--- a/arch/arm64/boot/dts/marvell/ac5_rd.dts
+++ b/arch/arm64/boot/dts/marvell/ac5_rd.dts
@@ -18,6 +18,10 @@
 	status = "okay";
 };
 
+&usb0 {
+	compatible = "marvell,ac5-ehci", "marvell,orion-ehci";
+};
+
 &usb1 {
 	compatible = "marvell,ac5-ehci", "marvell,orion-ehci";
 };
diff --git a/arch/arm64/configs/marvell_v8_sdk_defconfig b/arch/arm64/configs/marvell_v8_sdk_defconfig
index 4f892dda..eb464f9b 100644
--- a/arch/arm64/configs/marvell_v8_sdk_defconfig
+++ b/arch/arm64/configs/marvell_v8_sdk_defconfig
@@ -16,6 +16,7 @@ CONFIG_IKCONFIG=y
 CONFIG_IKCONFIG_PROC=y
 CONFIG_LOG_BUF_SHIFT=16
 CONFIG_LOG_CPU_MAX_BUF_SHIFT=13
+CONFIG_PRINTK_SAFE_LOG_BUF_SHIFT=13
 CONFIG_NUMA_BALANCING=y
 CONFIG_MEMCG=y
 CONFIG_MEMCG_SWAP=y
@@ -29,11 +30,6 @@ CONFIG_CGROUP_PERF=y
 CONFIG_USER_NS=y
 CONFIG_SCHED_AUTOGROUP=y
 CONFIG_BLK_DEV_INITRD=y
-# CONFIG_RD_BZIP2 is not set
-# CONFIG_RD_LZMA is not set
-# CONFIG_RD_XZ is not set
-# CONFIG_RD_LZO is not set
-# CONFIG_RD_LZ4 is not set
 CONFIG_KALLSYMS_ALL=y
 # CONFIG_COMPAT_BRK is not set
 CONFIG_PROFILING=y
@@ -43,16 +39,23 @@ CONFIG_MODULES=y
 CONFIG_MODULE_UNLOAD=y
 # CONFIG_IOSCHED_DEADLINE is not set
 CONFIG_ARCH_MVEBU=y
+CONFIG_ARCH_SPRD=y
+CONFIG_ARCH_THUNDER=y
 CONFIG_ARCH_VEXPRESS=y
 CONFIG_PCI=y
 CONFIG_HOTPLUG_PCI_PCIE=y
 CONFIG_PCI_IOV=y
 CONFIG_HOTPLUG_PCI=y
 CONFIG_HOTPLUG_PCI_ACPI=y
+CONFIG_PCI_HISI=y
 CONFIG_PCIE_ARMADA_8K=y
+CONFIG_PCIE_KIRIN=y
+CONFIG_PCI_AARDVARK=y
 CONFIG_PCI_HOST_GENERIC=y
+CONFIG_PCI_XGENE=y
 CONFIG_PCI_HOST_THUNDER_PEM=y
-# CONFIG_PCI_HOST_OCTEONTX_ECAM is not set
+CONFIG_PCI_HOST_OCTEONTX2_PEM=y
+CONFIG_PCI_HOST_THUNDER_ECAM=y
 CONFIG_PCI_ENDPOINT=y
 CONFIG_PCIE_ARMADA_DW_EP=y
 CONFIG_ARM64_VA_BITS_48=y
@@ -68,10 +71,9 @@ CONFIG_KEXEC=y
 CONFIG_CRASH_DUMP=y
 CONFIG_XEN=y
 CONFIG_ARM64_LSE_ATOMICS=y
-# CONFIG_ARM64_VHE is not set
 CONFIG_ARM64_PMEM=y
-# CONFIG_EFI is not set
 # CONFIG_CORE_DUMP_DEFAULT_ELF_HEADERS is not set
+CONFIG_HIBERNATION=y
 CONFIG_WQ_POWER_EFFICIENT_DEFAULT=y
 CONFIG_CPU_IDLE_GOV_MENU=y
 CONFIG_ARM_CPUIDLE=y
@@ -112,6 +114,7 @@ CONFIG_NET_MPLS_GSO=y
 CONFIG_BPF_JIT=y
 CONFIG_CFG80211=m
 CONFIG_MAC80211=m
+CONFIG_RFKILL=y
 CONFIG_NET_9P=y
 CONFIG_NET_9P_VIRTIO=y
 CONFIG_UEVENT_HELPER_PATH="/sbin/hotplug"
@@ -135,6 +138,7 @@ CONFIG_BLK_DEV_LOOP=y
 CONFIG_BLK_DEV_NBD=m
 CONFIG_BLK_DEV_RAM=y
 CONFIG_VIRTIO_BLK=y
+CONFIG_BLK_DEV_NVME=y
 CONFIG_SRAM=y
 CONFIG_EEPROM_AT24=y
 # CONFIG_SCSI_PROC_FS is not set
@@ -162,36 +166,28 @@ CONFIG_SATA_SIS=y
 CONFIG_PATA_MARVELL=y
 CONFIG_PATA_PLATFORM=y
 CONFIG_PATA_OF_PLATFORM=y
+CONFIG_MD=y
+CONFIG_BLK_DEV_MD=y
+CONFIG_MD_LINEAR=y
+CONFIG_MD_RAID0=y
+CONFIG_MD_RAID1=y
+CONFIG_MD_RAID456=y
+CONFIG_BLK_DEV_DM=y
+CONFIG_DM_CRYPT=y
 CONFIG_NETDEVICES=y
 CONFIG_BONDING=m
 CONFIG_MACVLAN=y
 CONFIG_MACVTAP=y
 CONFIG_TUN=y
 CONFIG_VETH=m
+CONFIG_VIRTIO_NET=y
 CONFIG_NET_DSA_MV88E6XXX=y
-# CONFIG_NET_VENDOR_3COM is not set
-# CONFIG_NET_VENDOR_ADAPTEC is not set
-# CONFIG_NET_VENDOR_AGERE is not set
-# CONFIG_NET_VENDOR_ALACRITECH is not set
-# CONFIG_NET_VENDOR_ALTEON is not set
-# CONFIG_NET_VENDOR_AMAZON is not set
-# CONFIG_NET_VENDOR_AMD is not set
-# CONFIG_NET_VENDOR_AQUANTIA is not set
-# CONFIG_NET_VENDOR_ARC is not set
-# CONFIG_NET_VENDOR_ATHEROS is not set
-# CONFIG_NET_CADENCE is not set
-# CONFIG_NET_VENDOR_BROADCOM is not set
-# CONFIG_NET_VENDOR_BROCADE is not set
-# CONFIG_NET_VENDOR_CAVIUM is not set
-# CONFIG_NET_VENDOR_CHELSIO is not set
-# CONFIG_NET_VENDOR_CISCO is not set
-# CONFIG_NET_VENDOR_DEC is not set
-# CONFIG_NET_VENDOR_EMULEX is not set
-# CONFIG_NET_VENDOR_EZCHIP is not set
-# CONFIG_NET_VENDOR_EXAR is not set
-# CONFIG_NET_VENDOR_HISILICON is not set
-# CONFIG_NET_VENDOR_HP is not set
-# CONFIG_NET_VENDOR_HUAWEI is not set
+CONFIG_AMD_XGBE=y
+CONFIG_MACB=y
+CONFIG_THUNDER_NIC_PF=y
+CONFIG_THUNDER_NIC_VF=y
+CONFIG_HNS_DSAF=y
+CONFIG_HNS_ENET=y
 CONFIG_E100=y
 CONFIG_E1000=y
 CONFIG_E1000E=y
@@ -205,38 +201,23 @@ CONFIG_I40EVF=y
 CONFIG_FM10K=y
 CONFIG_MVNETA=y
 CONFIG_MVPP2=y
-CONFIG_MVPP2X=y
+CONFIG_MVPP2X=m
+CONFIG_SKY2=y
 CONFIG_OCTEONTX2_AF=y
 CONFIG_OCTEONTX2_PF=y
 CONFIG_OCTEONTX2_VF=y
-# CONFIG_NET_VENDOR_MELLANOX is not set
-# CONFIG_NET_VENDOR_MICREL is not set
-# CONFIG_NET_VENDOR_MICROCHIP is not set
-# CONFIG_NET_VENDOR_MYRI is not set
-# CONFIG_NET_VENDOR_NATSEMI is not set
-# CONFIG_NET_VENDOR_NETRONOME is not set
-# CONFIG_NET_VENDOR_NVIDIA is not set
-# CONFIG_NET_VENDOR_OKI is not set
-# CONFIG_NET_PACKET_ENGINE is not set
-# CONFIG_NET_VENDOR_QLOGIC is not set
-# CONFIG_NET_VENDOR_QUALCOMM is not set
-# CONFIG_NET_VENDOR_REALTEK is not set
-# CONFIG_NET_VENDOR_RENESAS is not set
-# CONFIG_NET_VENDOR_RDC is not set
-# CONFIG_NET_VENDOR_ROCKER is not set
-# CONFIG_NET_VENDOR_SAMSUNG is not set
-# CONFIG_NET_VENDOR_SEEQ is not set
-# CONFIG_NET_VENDOR_SILAN is not set
-# CONFIG_NET_VENDOR_SIS is not set
-# CONFIG_NET_VENDOR_SOLARFLARE is not set
-# CONFIG_NET_VENDOR_SMSC is not set
-# CONFIG_NET_VENDOR_STMICRO is not set
-# CONFIG_NET_VENDOR_SUN is not set
-# CONFIG_NET_VENDOR_TEHUTI is not set
-# CONFIG_NET_VENDOR_TI is not set
-# CONFIG_NET_VENDOR_VIA is not set
-# CONFIG_NET_VENDOR_WIZNET is not set
-# CONFIG_NET_VENDOR_SYNOPSYS is not set
+CONFIG_MLX4_EN=y
+CONFIG_MLX5_CORE=y
+CONFIG_MLXSW_CORE=y
+CONFIG_8139CP=y
+CONFIG_8139TOO=y
+CONFIG_8139TOO_TUNE_TWISTER=y
+CONFIG_8139TOO_8129=y
+CONFIG_R8169=y
+CONFIG_SMC91X=y
+CONFIG_SMSC911X=y
+CONFIG_STMMAC_ETH=m
+CONFIG_MDIO_BITBANG=y
 CONFIG_MDIO_BUS_MUX_MMIOREG=y
 CONFIG_AQUANTIA_PHY=y
 CONFIG_BCM7XXX_PHY=y
@@ -258,7 +239,7 @@ CONFIG_PPP_ASYNC=m
 CONFIG_PPP_SYNC_TTY=m
 CONFIG_USB_PEGASUS=m
 CONFIG_USB_RTL8150=m
-CONFIG_USB_RTL8152=y
+CONFIG_USB_RTL8152=m
 CONFIG_USB_USBNET=y
 # CONFIG_USB_NET_CDCETHER is not set
 # CONFIG_USB_NET_CDC_NCM is not set
@@ -270,13 +251,13 @@ CONFIG_USB_NET_PLUSB=y
 CONFIG_USB_NET_MCS7830=y
 # CONFIG_USB_NET_CDC_SUBSET is not set
 # CONFIG_USB_NET_ZAURUS is not set
-# CONFIG_WLAN is not set
-# CONFIG_XEN_NETDEV_FRONTEND is not set
-CONFIG_INPUT_MATRIXKMAP=y
+CONFIG_WL18XX=m
+CONFIG_WLCORE_SDIO=m
 CONFIG_INPUT_MOUSEDEV=y
 CONFIG_INPUT_MOUSEDEV_PSAUX=y
 CONFIG_INPUT_EVDEV=y
 CONFIG_KEYBOARD_GPIO=y
+CONFIG_KEYBOARD_CROS_EC=y
 CONFIG_INPUT_MISC=y
 # CONFIG_SERIO_SERPORT is not set
 CONFIG_SERIO_AMBAKMI=y
@@ -306,6 +287,7 @@ CONFIG_I2C_MV64XXX=y
 CONFIG_I2C_PXA=y
 CONFIG_I2C_RK3X=y
 CONFIG_I2C_THUNDERX=y
+CONFIG_I2C_CROS_EC_TUNNEL=y
 CONFIG_I2C_SLAVE=y
 CONFIG_SPI=y
 CONFIG_SPI_ARMADA_3700=y
@@ -315,16 +297,19 @@ CONFIG_SPI_PL022=y
 CONFIG_SPI_ROCKCHIP=y
 CONFIG_SPI_THUNDERX=y
 CONFIG_SPI_SPIDEV=m
-CONFIG_PTP_1588_CLOCK_CAVIUM_THUNDERX_PTP=y
+CONFIG_SPMI=y
 CONFIG_PINCTRL_MCP23S08=y
 CONFIG_PINCTRL_SINGLE=y
+CONFIG_PINCTRL_MAX77620=y
 CONFIG_GPIO_SYSFS=y
 CONFIG_GPIO_DWAPB=y
 CONFIG_GPIO_PL061=y
+CONFIG_GPIO_THUNDERX=y
 CONFIG_GPIO_XGENE=y
 CONFIG_GPIO_I2C=y
 CONFIG_GPIO_PCA953X=y
 CONFIG_GPIO_PCA953X_IRQ=y
+CONFIG_GPIO_MAX77620=y
 CONFIG_POWER_RESET_BRCMSTB=y
 CONFIG_POWER_RESET_XGENE=y
 CONFIG_POWER_RESET_SYSCON=y
@@ -341,41 +326,59 @@ CONFIG_ARMADA_THERMAL=y
 CONFIG_WATCHDOG=y
 CONFIG_ARM_SBSA_WATCHDOG=y
 CONFIG_ARMADA3700_WATCHDOG=y
+CONFIG_MFD_CROS_EC=y
+CONFIG_MFD_CROS_EC_I2C=y
+CONFIG_MFD_CROS_EC_SPI=y
+CONFIG_MFD_HI6421_PMIC=y
+CONFIG_MFD_MAX77620=y
+CONFIG_MFD_RK808=y
+CONFIG_MFD_SEC_CORE=y
+CONFIG_REGULATOR=y
+CONFIG_REGULATOR_FIXED_VOLTAGE=y
+CONFIG_REGULATOR_FAN53555=y
+CONFIG_REGULATOR_GPIO=y
+CONFIG_REGULATOR_HI6421V530=y
+CONFIG_REGULATOR_MAX77620=y
+CONFIG_REGULATOR_PWM=y
+CONFIG_REGULATOR_QCOM_SPMI=y
+CONFIG_REGULATOR_RK808=y
+CONFIG_REGULATOR_S2MPS11=y
 # CONFIG_RC_CORE is not set
 CONFIG_DRM=m
 CONFIG_DRM_I2C_CH7006=m
 CONFIG_DRM_I2C_SIL164=m
 CONFIG_DRM_I2C_ADV7511=m
-# CONFIG_DRM_I2C_ADV7533 is not set
 CONFIG_DRM_HISI_KIRIN=m
 CONFIG_FB=y
-CONFIG_FB_MODE_HELPERS=y
-# CONFIG_XEN_FBDEV_FRONTEND is not set
-# CONFIG_HID_GENERIC is not set
+CONFIG_FB_ARMCLCD=y
+# CONFIG_LCD_CLASS_DEVICE is not set
+# CONFIG_BACKLIGHT_GENERIC is not set
+CONFIG_BACKLIGHT_LP855X=m
+CONFIG_LOGO=y
+# CONFIG_LOGO_LINUX_MONO is not set
+# CONFIG_LOGO_LINUX_VGA16 is not set
 CONFIG_USB=y
 CONFIG_USB_OTG=y
 CONFIG_USB_XHCI_HCD=y
 CONFIG_USB_EHCI_HCD=y
-CONFIG_USB_EHCI_ROOT_HUB_TT=y
 CONFIG_USB_EHCI_HCD_PLATFORM=y
 CONFIG_USB_OHCI_HCD=y
 CONFIG_USB_OHCI_HCD_PLATFORM=y
 CONFIG_USB_STORAGE=y
 CONFIG_USB_DWC3=y
-CONFIG_USB_DWC3_HOST=y
 CONFIG_USB_DWC2=y
-CONFIG_USB_DWC2_HOST=y
 CONFIG_USB_CHIPIDEA=y
 CONFIG_USB_CHIPIDEA_UDC=y
+CONFIG_USB_CHIPIDEA_HOST=y
+CONFIG_USB_ISP1760=y
+CONFIG_USB_HSIC_USB3503=y
 CONFIG_NOP_USB_XCEIV=y
 CONFIG_USB_ULPI=y
 CONFIG_USB_GADGET=y
-CONFIG_USB_CONFIGFS=y
-CONFIG_USB_CONFIGFS_SERIAL=y
-CONFIG_USB_CONFIGFS_ACM=y
-CONFIG_USB_CONFIGFS_OBEX=y
-CONFIG_USB_CONFIGFS_MASS_STORAGE=y
-CONFIG_USB_G_SERIAL=y
+CONFIG_USB_MV_UDC=y
+CONFIG_USB_MV_U3D=y
+CONFIG_USB_SNP_UDC_PLAT=y
+CONFIG_USB_BDC_UDC=y
 CONFIG_MMC=y
 CONFIG_MMC_BLOCK_MINORS=32
 CONFIG_MMC_OOPS=y
@@ -395,13 +398,19 @@ CONFIG_EDAC=y
 CONFIG_EDAC_ARMADA_AP80X=y
 CONFIG_RTC_CLASS=y
 CONFIG_RTC_DRV_DS1307=y
+CONFIG_RTC_DRV_MAX77686=y
+CONFIG_RTC_DRV_S5M=y
 CONFIG_RTC_DRV_DS3232=y
+CONFIG_RTC_DRV_EFI=y
 CONFIG_RTC_DRV_PL031=y
 CONFIG_RTC_DRV_ARMADA38X=y
 CONFIG_DMADEVICES=y
 CONFIG_MV_XOR=y
 CONFIG_MV_XOR_V2=y
 CONFIG_PL330_DMA=y
+CONFIG_QCOM_HIDMA_MGMT=y
+CONFIG_QCOM_HIDMA=y
+CONFIG_ASYNC_TX_DMA=y
 CONFIG_UIO=y
 CONFIG_UIO_PDRV_GENIRQ=m
 CONFIG_UIO_PCI_GENERIC=m
@@ -410,30 +419,29 @@ CONFIG_VFIO=y
 CONFIG_VFIO_PCI=y
 CONFIG_VFIO_PLATFORM=y
 CONFIG_VFIO_PLATFORM_XHCI_RESET=y
+CONFIG_VIRT_DRIVERS=y
 CONFIG_VIRTIO_PCI=y
 CONFIG_VIRTIO_BALLOON=y
 CONFIG_VIRTIO_MMIO=y
-# CONFIG_XEN_BALLOON is not set
-# CONFIG_XEN_DEV_EVTCHN is not set
-# CONFIG_XEN_BACKEND is not set
-# CONFIG_XENFS is not set
-# CONFIG_XEN_SYS_HYPERVISOR is not set
-# CONFIG_XEN_GNTDEV is not set
-# CONFIG_XEN_GRANT_DEV_ALLOC is not set
+CONFIG_XEN_GNTDEV=y
+CONFIG_XEN_GRANT_DEV_ALLOC=y
 CONFIG_STAGING=y
+CONFIG_COMMON_CLK_RK808=y
 CONFIG_COMMON_CLK_SCMI=y
 CONFIG_COMMON_CLK_SCPI=y
 CONFIG_COMMON_CLK_CS2000_CP=y
+CONFIG_COMMON_CLK_S2MPS11=y
 CONFIG_CLK_QORIQ=y
+CONFIG_COMMON_CLK_PWM=y
 CONFIG_HWSPINLOCK=y
 # CONFIG_ARM_ARCH_TIMER_EVTSTREAM is not set
-# CONFIG_FSL_ERRATUM_A008585 is not set
-# CONFIG_HISILICON_ERRATUM_161010101 is not set
-# CONFIG_ARM64_ERRATUM_858921 is not set
+CONFIG_ARM_TIMER_SP804=y
 CONFIG_MAILBOX=y
 CONFIG_ARM_MHU=y
+CONFIG_MVL_MHU=y
 CONFIG_PLATFORM_MHU=y
 CONFIG_PCC=y
+CONFIG_BCM_FLEXRM_MBOX=y
 CONFIG_ARM_SMMU=y
 CONFIG_ARM_SMMU_V3=y
 CONFIG_OCTEONTX2_RM=y
@@ -442,25 +450,31 @@ CONFIG_OCTEONTX2_DPI_PF=y
 CONFIG_GTI_WATCHDOG=y
 CONFIG_OCTEONTX_INFO=y
 CONFIG_EXTCON_USB_GPIO=y
+CONFIG_IIO=y
+CONFIG_PWM=y
 CONFIG_PHY_XGENE=y
 CONFIG_PHY_MVEBU_CP110_COMPHY=y
 CONFIG_PHY_SAMSUNG_USB2=y
-# CONFIG_PHONE is not set
+CONFIG_MVEBU_PHONE=y
 CONFIG_TEE=y
+CONFIG_OPTEE=y
 CONFIG_ARM_SCMI_PROTOCOL=y
 # CONFIG_ARM_SCMI_POWER_DOMAIN is not set
 CONFIG_ARM_SCPI_PROTOCOL=y
+CONFIG_EFI_CAPSULE_LOADER=y
 CONFIG_ACPI=y
-CONFIG_ACPI_HED=y
-CONFIG_EXT2_FS=m
+CONFIG_ACPI_APEI=y
+CONFIG_ACPI_APEI_GHES=y
+CONFIG_ACPI_APEI_PCIEAER=y
+CONFIG_EXT2_FS=y
 CONFIG_EXT3_FS=y
 CONFIG_EXT4_FS_POSIX_ACL=y
+CONFIG_XFS_FS=y
 CONFIG_BTRFS_FS=y
 CONFIG_BTRFS_FS_POSIX_ACL=y
 CONFIG_FANOTIFY=y
 CONFIG_FANOTIFY_ACCESS_PERMISSIONS=y
 CONFIG_QUOTA=y
-# CONFIG_PRINT_QUOTA_WARNING is not set
 CONFIG_AUTOFS4_FS=y
 CONFIG_FUSE_FS=m
 CONFIG_CUSE=m
@@ -468,12 +482,16 @@ CONFIG_OVERLAY_FS=m
 CONFIG_VFAT_FS=y
 CONFIG_TMPFS=y
 CONFIG_HUGETLBFS=y
+CONFIG_CONFIGFS_FS=y
+CONFIG_EFIVAR_FS=y
+CONFIG_JFFS2_FS=y
 CONFIG_UBIFS_FS=y
 CONFIG_UBIFS_FS_ADVANCED_COMPR=y
 CONFIG_SQUASHFS=y
 CONFIG_SQUASHFS_LZO=y
 CONFIG_PSTORE=y
 CONFIG_PSTORE_RAM=y
+CONFIG_PANIC_TIMEOUT=1
 CONFIG_NFS_FS=y
 CONFIG_NFS_V4=y
 CONFIG_NFS_V4_1=y
@@ -492,7 +510,6 @@ CONFIG_DEBUG_FS=y
 CONFIG_MAGIC_SYSRQ=y
 CONFIG_DEBUG_KERNEL=y
 CONFIG_DETECT_HUNG_TASK=y
-CONFIG_PANIC_TIMEOUT=1
 # CONFIG_SCHED_DEBUG is not set
 # CONFIG_DEBUG_PREEMPT is not set
 # CONFIG_FTRACE is not set
@@ -508,7 +525,6 @@ CONFIG_CRYPTO_TEST=m
 CONFIG_CRYPTO_CCM=y
 CONFIG_CRYPTO_GCM=y
 CONFIG_CRYPTO_ECHAINIV=y
-CONFIG_CRYPTO_CBC=y
 CONFIG_CRYPTO_CTS=y
 CONFIG_CRYPTO_CMAC=y
 CONFIG_CRYPTO_MD5=y
@@ -520,6 +536,8 @@ CONFIG_CRYPTO_USER_API_HASH=m
 CONFIG_CRYPTO_USER_API_SKCIPHER=m
 CONFIG_CRYPTO_USER_API_RNG=m
 CONFIG_CRYPTO_USER_API_AEAD=m
+CONFIG_CRYPTO_DEV_OCTEONTX_CPT=y
+CONFIG_CRYPTO_DEV_OCTEONTX2_CPT=y
 CONFIG_CRYPTO_DEV_SAFEXCEL=m
 CONFIG_ARM64_CRYPTO=y
 CONFIG_CRYPTO_SHA512_ARM64=y
@@ -532,5 +550,4 @@ CONFIG_CRYPTO_AES_ARM64_CE_CCM=y
 CONFIG_CRYPTO_AES_ARM64_CE_BLK=y
 CONFIG_CRYPTO_CHACHA20_NEON=y
 CONFIG_CRYPTO_AES_ARM64_BS=y
-CONFIG_LIBCRC32C=y
-CONFIG_XZ_DEC=y
+CONFIG_SENSORS_LM75=y
diff --git a/drivers/pci/pci.c b/drivers/pci/pci.c
index 1af30c88..c5c469e0 100644
--- a/drivers/pci/pci.c
+++ b/drivers/pci/pci.c
@@ -1570,6 +1570,7 @@ EXPORT_SYMBOL(pcim_pin_device);
  */
 int __weak pcibios_add_device(struct pci_dev *dev)
 {
+	dev->irq = of_irq_parse_and_map_pci(dev, 0, 0);
 	return 0;
 }
 
diff --git a/drivers/pci/pcie/portdrv_core.c b/drivers/pci/pcie/portdrv_core.c
index ebc9d45b..d5663c62 100644
--- a/drivers/pci/pcie/portdrv_core.c
+++ b/drivers/pci/pcie/portdrv_core.c
@@ -64,7 +64,7 @@ static int pcie_port_enable_irq_vec(struct pci_dev *dev, int *irqs, int mask)
 	 * go through without any tricks.
 	 */
 	nr_entries = pci_alloc_irq_vectors(dev, 1, PCIE_PORT_MAX_MSI_ENTRIES,
-			PCI_IRQ_MSIX | PCI_IRQ_MSI);
+			PCI_IRQ_MSI);
 	if (nr_entries < 0)
 		return nr_entries;
 
@@ -89,7 +89,7 @@ static int pcie_port_enable_irq_vec(struct pci_dev *dev, int *irqs, int mask)
 		pcie_capability_read_word(dev, PCI_EXP_FLAGS, &reg16);
 		entry = (reg16 & PCI_EXP_FLAGS_IRQ) >> 9;
 		if (entry >= nr_entries)
-			goto out_free_irqs;
+			return 0;
 
 		irqs[PCIE_PORT_SERVICE_PME_SHIFT] = pci_irq_vector(dev, entry);
 		irqs[PCIE_PORT_SERVICE_HP_SHIFT] = pci_irq_vector(dev, entry);
diff --git a/drivers/usb/host/ehci-orion.c b/drivers/usb/host/ehci-orion.c
index 3d353043..7ffdbe03 100644
--- a/drivers/usb/host/ehci-orion.c
+++ b/drivers/usb/host/ehci-orion.c
@@ -160,6 +160,21 @@ ehci_orion_conf_mbus_windows(struct usb_hcd *hcd,
 	}
 }
 
+/* Configure USB's DMA window to be lower part of DRAM @ address 0 */
+static void
+ehci_orion_conf_ac5_windows(struct usb_hcd *hcd)
+{
+#define USB_TO_DRAM_TARGET_ID 0x2
+#define USB_TO_DRAM_ATTR_ID 0x0
+#define USB_DRAM_BASE 0x00000000
+#define USB_DRAM_SIZE 0xffff0000
+
+	wrl(USB_WINDOW_CTRL(0),
+		(USB_DRAM_SIZE & 0xffff0000) |
+		(USB_TO_DRAM_ATTR_ID << 8) | (USB_TO_DRAM_TARGET_ID << 4) | 1);
+	wrl(USB_WINDOW_BASE(0), USB_DRAM_BASE);
+}
+
 static int ehci_orion_drv_reset(struct usb_hcd *hcd)
 {
 	struct device *dev = hcd->self.controller;
@@ -200,6 +215,8 @@ static int ehci_orion_drv_probe(struct platform_device *pdev)
 	int irq, err;
 	enum orion_ehci_phy_ver phy_version;
 	struct orion_ehci_hcd *priv;
+	struct device_node *dn = pdev->dev.of_node;
+	u32 val;
 
 	if (usb_disabled())
 		return -ENODEV;
@@ -277,6 +294,13 @@ static int ehci_orion_drv_probe(struct platform_device *pdev)
 	if (dram)
 		ehci_orion_conf_mbus_windows(hcd, dram);
 
+	if (of_device_is_compatible(pdev->dev.of_node, "marvell,ac5-ehci")) {
+		ehci_orion_conf_ac5_windows(hcd);
+		/* Declare, that although device is 32bit address space DMA,
+		   its DMA address space starts at 0x200,000,000 */
+		pdev->dev.dma_pfn_offset = 2 << (32 - PAGE_SHIFT);
+	}
+
 	/*
 	 * setup Orion USB controller.
 	 */
@@ -300,6 +324,12 @@ static int ehci_orion_drv_probe(struct platform_device *pdev)
 	err = usb_add_hcd(hcd, irq, IRQF_SHARED);
 	if (err)
 		goto err_add_hcd;
+	if (of_device_is_compatible(pdev->dev.of_node, "marvell,ac5-ehci")) {
+		if (of_get_property(dn, "max-speed-full-speed", NULL)) {
+			val = rdl(USB_PORTSC_REG_OFF) | (1 << 24);
+			wrl(USB_PORTSC_REG_OFF, val);
+		}
+	}
 
 	device_wakeup_enable(hcd->self.controller);
 	return 0;
@@ -345,6 +375,7 @@ static int ehci_orion_drv_remove(struct platform_device *pdev)
 static const struct of_device_id ehci_orion_dt_ids[] = {
 	{ .compatible = "marvell,orion-ehci", },
 	{ .compatible = "marvell,armada-3700-ehci", },
+	{ .compatible = "marvell,ac5-ehci", },
 	{ },
 };
 MODULE_DEVICE_TABLE(of, ehci_orion_dt_ids);
-- 
2.25.1

